crates/matrix-sdk-crypto/src/identities/user.rs

```diff
@@ -1266,18 +1274,126 @@ pub(crate) mod testing {
+    ///         }
+    ///         "usage": [ "master" ],
+    ///         "user_id": "their_user_id"
+    ///     }
+    /// });
+    ///
+    /// let ssk_json = json!({
+    ///     "their_user_id": {
+    ///         "keys": { "ed25519:blah": "blah" },
+    ///         "signatures": {
+    ///             "their_user_id": { "ed25519:blah": "blah" }
+    ///         },
+    ///         "usage": [ "self_signing" ],
+    ///         "user_id": "their_user_id"
+    ///     }
+    /// })
+    ///
+    /// let response = simulate_key_query_response_for_verification(
+    ///     signature_upload_request,
+    ///     my_identity,
+    ///     my_user_id,
+    ///     their_user_id,
+    ///     msk_json,
+    ///     ssk_json
+    /// ).await;
+    ///
+    /// olm_machine
+    ///     .mark_request_as_sent(
+    ///         &TransactionId::new(),
+    ///         crate::IncomingResponse::KeysQuery(&kq_response),
+    ///     )
+    ///     .await
+    ///     .unwrap();
+    /// ```
+    pub fn simulate_key_query_response_for_verification(
+        signature_upload_request: SignatureUploadRequest,
+        my_identity: OwnUserIdentity,
+        my_user_id: &UserId,
+        their_user_id: &UserId,
+        msk_json: serde_json::Value,
+        ssk_json: serde_json::Value,
+    ) -> KeyQueryResponse {
+        // Find the signed key inside the SignatureUploadRequest
+        let cross_signing_key: CrossSigningKey = serde_json::from_str(
+            signature_upload_request
+                .signed_keys
+                .get(their_user_id)
+                .expect("Signature upload request should contain a key for their user ID")
+                .iter()
+                .next()
+                .expect("There should be a key in the signature upload request")
+                .1
+                .get(),
+        )
+        .expect("Should not fail to deserialize the key");
+
+        // Find their master key that we want to update inside their msk JSON
+        let mut their_msk: CrossSigningKey = serde_json::from_value(
+            msk_json.get(their_user_id.as_str()).expect("msk should contain their user ID").clone(),
+        )
+        .expect("Should not fail to deserialize msk");
+
+        // Find our own user signing key
+        let my_user_signing_key_id = my_identity
+            .user_signing_key()
+            .keys()
+            .iter()
+            .next()
+            .expect("There should be a user signing key")
+            .0;
+
+        // Add the signature from the SignatureUploadRequest to their master key, under
+        // our user ID
+        their_msk.signatures.add_signature(
+            my_user_id.to_owned(),
+            my_user_signing_key_id.to_owned(),
+            cross_signing_key
+                .signatures
+                .get_signature(my_user_id, my_user_signing_key_id)
+                .expect("There should be a signature for our user"),
+        );
```

Comment:
Another thought on this: a much more intuitive way to generate this signature is to build it ourselves, rather than abusing `OtherUserIdentity::verify` (which of course requires the offending identity to have been imported into the olm machine).

Building the signature ourselves requires getting hold of our own private USK, but that's not so bad:

```rust
    async fn get_private_usk(machine: &OlmMachine) -> Ed25519SecretKey {
        Ed25519SecretKey::from_base64(
            machine
                .export_cross_signing_keys()
                .await
                .unwrap()
                .unwrap()
                .user_signing_key
                .as_ref()
                .unwrap(),
        )
        .unwrap()
    }
```

And then signing the MSK is something like

```
        use crate::olm::utility::SignJson;
        let signature = my_usk.sign_json(serde_json::to_value(&msk_to_update).unwrap()).unwrap(),
        let my_user_signing_key_id = DeviceKeyId::from_parts(
                DeviceKeyAlgorithm::Ed25519,
                my_usk.public_key().to_base64().as_str().into(),
        );

        msk_to_update.signatures.add_signature(
            my_user_id.to_owned(),
            my_user_signing_key_id,
            signature
        );
```

The problem with that is that `crate::olm::utility::SignJson` is `pub(crate)`, but maybe we could add a `#[cfg(test)]`?

Reply:
> The problem with that is that crate::olm::utility::SignJson is pub(crate), but maybe we could add a #[cfg(test)]?

Why is that a problem this PR is completely inside the crypto crate?

Reply:
> Why is that a problem this PR is completely inside the crypto crate?

Sorry, I'm confused. It's private, not pub(crate).

`utility::{SignedJsonObject, VerifyJson}` are exported from `olm` at pub(crate); maybe `SignJson` should be too.

Reply:
This seems like a good plan but I don't have time to work on it now. Shall we make an issue for it, or just leave the info here?

Reply:
> Sorry, I'm confused. It's private, not pub(crate).

I think I left that private for a reason and implemented higher level methods like `sign_device()` on things that can sign devices. It is supposed to encode in our type system which things are allowed to sign which objects.

Exposing the general signing mechanism and re-implementing `sign_user()` in a test helper doesn't sound that great to me. Refactoring `sign_user()` into  two methods, one which returns a request, and another one which returns a `CrossSigningKey` sounds better to me.

Reply:
> This seems like a good plan but I don't have time to work on it now. Shall we make an issue for it, or just leave the info here?

I can't see that an issue is likely to be very useful. It will just clutter things up and we won't remember to look for it when we come to do the work.

Reply:
> I don't really agree with this: I think that having to construct the higher-level objects makes testing unnecessarily difficult, and makes tests fragile, complicated, and generally hard to maintain.

That's a fair argument, but the example Rust code you posted requires the same higher-level objects to be created as my suggestion would.